service/topology_coordinator.cc

```diff
@@ -66,6 +67,757 @@ future<inet_address> wait_for_ip(raft::server_id id, const raft_address_map& am,
+    }
+}
+
+// ----------------------------------------------------------------------------
+    using global_shard_id = locator::tablet_replica;
+    using shard_id = seastar::shard_id;
+
+    // Represents metric for per-node load which we want to equalize between nodes.
+    // It's an average per-shard load in terms of tablet count.
+    using load_type = double;
+
+    struct shard_load {
+        size_t tablet_count = 0;
+
+        // Number of tablets which are streamed from this shard.
+        size_t streaming_read_load = 0;
+
+        // Number of tablets which are streamed to this shard.
+        size_t streaming_write_load = 0;
+
+        // Tablets which still have a replica on this shard which are candidates for migrating away from this shard.
+        std::unordered_set<locator::global_tablet_id> candidates;
+
+        future<> clear_gently() {
+            return utils::clear_gently(candidates);
+        }
+    };
+
+    struct node_load {
+        locator::host_id id;
+        uint64_t shard_count = 0;
+        uint64_t tablet_count = 0;
+
+        // The average shard load on this node.
+        load_type avg_load = 0;
+
+        std::vector<shard_id> shards_by_load; // heap which tracks most-loaded shards using shards_by_load_cmp().
+        std::vector<shard_load> shards; // Indexed by shard_id to which a given shard_load corresponds.
+
+        std::optional<locator::load_sketch> target_load_sketch;
+
+        future<locator::load_sketch&> get_load_sketch(const locator::token_metadata_ptr& tm) {
+            if (!target_load_sketch) {
+                target_load_sketch.emplace(tm);
+                co_await target_load_sketch->populate(id);
+            }
+            co_return *target_load_sketch;
+        }
+
+        // Call when tablet_count changes.
+        void update() {
+            avg_load = get_avg_load(tablet_count);
+        }
+
+        load_type get_avg_load(uint64_t tablets) const {
+            return double(tablets) / shard_count;
+        }
+
+        auto shards_by_load_cmp() {
+            return [this] (const auto& a, const auto& b) {
+                return shards[a].tablet_count < shards[b].tablet_count;
+            };
+        }
+
+        future<> clear_gently() {
+            return utils::clear_gently(shards);
+        }
+    };
+
+struct load_balancer_dc_stats {
+    uint64_t calls = 0;
+    uint64_t migrations_produced = 0;
+    uint64_t migrations_skipped = 0;
+    uint64_t tablets_skipped_node = 0;
+    uint64_t tablets_skipped_rack = 0;
+    uint64_t stop_balance = 0;
+    uint64_t stop_load_inversion = 0;
+    uint64_t stop_no_candidates = 0;
+    uint64_t stop_skip_limit = 0;
+    uint64_t stop_batch_size = 0;
+};
+
+struct load_balancer_node_stats {
+    double load = 0;
+};
+
+struct load_balancer_cluster_stats {
+    uint64_t resizes_emitted = 0;
+    uint64_t resizes_revoked = 0;
+    uint64_t resizes_finalized = 0;
+};
+
+class load_balancer_stats_manager {
```

Comment:
This looks like copy-paste of code from tablet_allocator.cc. I don't know what's the intention here, I don't know how to review this change.

Reply:
We'll split this PR into 2:
1. here we'll only leave the request processing path and stub `reallocate_tablets_for_new_rf`
2. in https://github.com/scylladb/scylladb/pull/17116 @KrzaQ will implement `reallocate_tablets_for_new_rf`, 

I'll remove the 2nd part from this PR and @KrzaQ will address this comment in https://github.com/scylladb/scylladb/pull/17116